### ISSUE TITLE
`no-this-before-super` rule false positive

### ISSUE THREAD

**Issue by overlookmotel** [eslint/eslint#13554]
**Tell us about your environment**

ESLint version: v7.6.0
Node version: v14.7.0
npm version: v6.14.7

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  extends: 'eslint:recommended',
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: 'script'
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
class X {}

class Y extends X {
  constructor() {
    const {a = this} = super();
    this.b = a;
  }
}
```

```bash
eslint .
```

**What did you expect to happen?**

`no-this-before-super` rule not to be triggered. This is valid (if unusual) code.

Other code where execution order is right-to-left may also trigger the rule incorrectly.

**What actually happened? Please include the actual, raw output from ESLint.**

```bash
5:14  error  'this' is not allowed before 'super()'  no-this-before-super
```

**Are you willing to submit a pull request to fix this bug?**

Sorry, no. It's a very unusual case, so probably shouldn't be a priority for anyone! I just raise this issue for completeness.

**Comment by mdjermanovic** [comment]
Hi @overlookmotel, thanks for the bug report!

I can confirm that this is a false positive, but added "evaluating" label since I'm not sure yet if we should fix it.

It looks like this would be a fairly complex fix. The problem is that our code path analysis follows the traversal order, which isn't always the same as the execution order.

In this particular case, it does produce a bug (false positive). On the other hand, as you noted it's a very unusual case, so I'm not sure would it justify the added complexity. I believe that any usage of `super()` in a non-statement position is uncommon.
